### DISCUSSION TITLE
External library access to user-defined Modelica functions

### DISCUSSION THREAD

**Discussion by tburch-LSS** [OpenModelica/OpenModelica#11848] [Q&A]
as a contributor to the GSL extension to Modelica, i have discovered that it would be extremely helpful for routines in an external library to have access to user-defined Modelica functions. (e.g., at the moment i am trying to incorporate GSL's integration routines, which would require access to any integrand which a user may define within, say, the OMEdit environment.)

as i see it, there are two possibilities: 1. the external library retains access to the flattened / translated C functions via some standardized function-naming conventions (and possibly additional interface routines added to the external library beforehand). or... 2. one adds a pre-model-compilation step which constructs new C routines from the Modelica functions, compiles them, and adds them to the external library (using Modelica.Utilities capabilities); then let the model compile and link with the newly augmented external library.

any ideas, thoughts, or similar experiences?

( i started a similar discussion over at the Modelica page:
https://github.com/modelica/ModelicaStandardLibrary/discussions/4230 )

**Comment by casella** [comment]
@fedetft any thoughts?

**Comment by tburch-LSS** [comment]
possibility 1 would probably be more elegant, but i'm leaning toward 2. so i'll refine my question thus:

is there a way to setup a model in OMEdit such that before the model is flattened (or at least before it compiles), a set of routines (from Modelica.Utilities) or an initialization script (*.mos) is run? (such a set / script could handle the translation / creation / compilation of C versions of any needed user-defined Modelica functions, which can then be called by the external library.)

**Comment by tburch-LSS** [comment]
for more on solution 2, see:
https://github.com/modelica/ModelicaStandardLibrary/discussions/4230#discussioncomment-10547555
